src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/DynamicMethod.cs

```diff
@@ -225,28 +226,25 @@ private static void CheckConsistency(MethodAttributes attributes, CallingConvent
         [System.Security.DynamicSecurityMethod] // Methods containing StackCrawlMark local var has to be marked DynamicSecurityMethod
         private static RuntimeModule GetDynamicMethodsModule()
         {
-            if (s_anonymouslyHostedDynamicMethodsModule != null)
-                return s_anonymouslyHostedDynamicMethodsModule;
-
-            lock (s_anonymouslyHostedDynamicMethodsModuleLock)
-            {
-                if (s_anonymouslyHostedDynamicMethodsModule != null)
-                    return s_anonymouslyHostedDynamicMethodsModule;
-
-                AssemblyName assemblyName = new AssemblyName("Anonymously Hosted DynamicMethods Assembly");
-                StackCrawlMark stackMark = StackCrawlMark.LookForMe;
-
-                AssemblyBuilder assembly = AssemblyBuilder.InternalDefineDynamicAssembly(
-                    assemblyName,
-                    AssemblyBuilderAccess.Run,
-                    ref stackMark,
-                    null);
-
-                // this always gets the internal module.
-                s_anonymouslyHostedDynamicMethodsModule = (InternalModuleBuilder)assembly.ManifestModule!;
-            }
-
-            return s_anonymouslyHostedDynamicMethodsModule;
+            return s_anonymouslyHostedDynamicMethodsModules.GetValue
+            (
+                AssemblyLoadContext.CurrentContextualReflectionContext ?? AssemblyLoadContext.Default,
+                assemblyLoadContext =>
+                {
+                    AssemblyName assemblyName = new AssemblyName("Anonymously Hosted DynamicMethods Assembly");
```

Comment:
What is the real world scenario fixed by creating multiple of these for dynamic methods?

Reply:
For example, creating a dynamic method in different assembly loading contexts.
In the old logic, there was only one dynamic assembly instance for storing dynamic methods, and the dynamic assembly was attached to default ALC.

Reply:
Right, but I do not see anything wrong with the existing behavior. Attaching all dynamic to the default context and the one-and-only Anonymous DynamicMethods assembly should work just fine.

Reply:
Not for the case when types from the unloaded assembly are used.

Reply:
Unlike dynamic assemblies, dynamic methods do not bind to anything. All references are provided pre-bound  when the dynamic method is created. It is why the dynamic method load context should not matter.

Reply:
Hmm .... Leave loading to the default context and roll back changes in DynamicMethod?

Reply:
I think so.

Reply:
Changes to DynamicMethod rolled back

Reply:
Just curious: how will this work with unloadability - how does it keep references to dependencies?

Reply:
DynamicMethods use reflection objects to reference dependencies. The reflection objects keep things alive as necessary. 

It is different from dynamic assemblies. Dynamic assemblies use regular metadata tokens for references.

Reply:
And what is the lifetime of a dynamic method - does it get freed (and all those references are released) when there are no references to the dynamic method itself? (since the loader it lives in as everlasting it seems).

Reply:
Yes, the dynamic assembly for the dynamic method is created with `AssemblyBuilderAccess.Run` => assembly loader is everlasting

Reply:
> what is the lifetime of a dynamic method - does it get freed (and all those references are released) when there are no references to the dynamic method itself?

Correct. Dynamic methods have its own lifetime tracking mechanism, different from loader allocators.

Reply:
Do I understand correctly that a dynamic method can be unloaded at any time, regardless of the context? What will happen to the assembly of a dynamic method?

Reply:
Yep